### ISSUE TITLE
[Proposal] Allow foreach for IEnumerator/IAsyncEnumerator

### ISSUE THREAD

**Issue by Kir-Antipov** [dotnet/csharplang#3867]
_For the sake of simplicity, I'll only talk about `IEnumerator`, but this issue also applies to the `IEnumerator<T>` and `IAsyncEnumerator<T>`_

# The problem

Let's imagine that we need to treat the first element of the enumeration differently than the rest. How can we do this today?

### 1. Condition inside the `foreach`-loop:
```csharp
bool isFirst = true;
foreach (var x in enumeration)
{
    if (isFirst)
    {
        isFirst = false;
        DoSomeWork(x);
    }
    else
    {
        DoAnotherWork(x);
    }
}
```
Pros:
 - No reevaluating

Cons:
 - Unnecessary branching for all elements of the enumeration

### 2. Reevaluating an enumeration:
```csharp
DoSomeWork(enumeration.First());
foreach (var x in enumeration.Skip(1))
    DoAnotherWork(x);
```
Pros:
 - It's short and simple

Cons:
 - We're reevaluating the enumeration and calculate its first element twice. This can be quite expensive, especially when it comes to asynchronous enumerations.

### 3. Extension method that wraps IEnumerator in new IEnumerable

```csharp
public static IEnumerable<T> ToEnumerable<T>(this IEnumerator<T> enumerator)
{
    _ = enumerator ?? throw new ArgumentNullException(nameof(enumerator));

    while (enumerator.MoveNext())
        yield return enumerator.Current;
}

...

using IEnumerator<object> e = enumeration.GetEnumerator();
DoSomeWork(e.ToEnumerable().First());
foreach (var x in e.ToEnumerable())
   DoAnotherWork(x);
```
Pros:
 - No reevaluating

Cons:
 - Unnecessary allocation
 - It's not built-in feature, so we must reinvent this from project to project
 - Really strange logic: turn an enumerator into an enumerable so that `foreach` can get an enumerator from it...

![confused-jackie-chan](https://user-images.githubusercontent.com/40232067/92625055-e35f1480-f2d0-11ea-8551-2c34c0c8d802.jpg)

### 4. Hello ~~darkness~~ boilerplate, my old friend:
```csharp
IEnumerator e = enumeration.GetEnumerator();
try
{
    if (!e.MoveNext())
        throw new InvalidOperationException()
    DoSomeWork(e.Current);
    while (e.MoveNext())
        DoAnotherWork(e.Current);
}
finally
{
    if (e is IDisposable disposable)
        disposable.Dispose();
}
```

Pros:
 - No cons of the previous methods

Cons:
 - It's duckling boring and nobody really has patience to write this

# Possible solution

`foreach` needs an `IEnumerator` to work, right? So why not let him work with it directly and without intermediaries? This would allow us to do things like this:

```csharp
using IEnumerator<int> e = new[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 }.Cast<int>().GetEnumerator();

// .TakeNext - LINQ-like extension method, that returns new enumerator,
// which can be implemented as a structure and be completely allocation-free 
foreach (int x in e.TakeNext(1))
    Console.WriteLine("First block:  {0}", x);

foreach (int x in e.TakeNext(2))
    Console.WriteLine("Second block: {0}", x);

foreach (int x in e.TakeNext(3))
    Console.WriteLine("Third block:  {0}", x);

foreach (int x in e)
    Console.WriteLine("Last block:   {0}", x);

// First block:  1
// Second block: 2
// Second block: 3
// Third block:  4
// Third block:  5
// Third block:  6
// Last block:   7
// Last block:   8
// Last block:   9
// Last block:   10
```

**Comment by HaloFour** [comment]
> `foreach` needs an `IEnumerator` to work, right?

No, `foreach` requires `GetEnumerator` that returns a type that fits the enumerator pattern.  You aren't required to implement the interfaces, just surface compatible methods.  You can already write an extension method for `IEnumerator<T>` today that does exactly what you're requesting.

**Comment by Kir-Antipov** [comment]
> You can already write an extension method for IEnumerator<T> today that does exactly what you're requesting

@HaloFour, yes indeed. And it'is completely different from what I'm talking about in **3**

**Comment by 333fred** [comment]
No, that is not what you talk about in 3. HaloFour means that, with C# 9, you can write this extension method:

```cs
public IEnumerator<T> GetEnumerator<T>(this IEnumerator<T> t) => t;
```

and it will just work :tm:

**Comment by HaloFour** [comment]
@Kir-Antipov 

> Unnecessary allocation

No allocations necessary, enumerables/enumerators can be structs and do not need to implement the interface.

> It's not built-in feature, so we must reinvent this from project to project

This is an uncommon use case and can be solved with a library method with little/no overhead and very little code.  There is very little reason to consider this as a language change.  You're not even proposing a language change, you're proposing a new API, one that you can write today in a dozen or so lines of code.  If it's that useful, publish it to NuGet.  I wouldn't be shocked if there is an implementation up there already.

> Really strange logic: turn an enumerator into an enumerable so that foreach can get an enumerator from it...

If it's strange then why are you asking to do it?  Turning an enumerator back into an enumerable is _exactly_ what you're asking for here.

[SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAgrgOwD4AEBMAjALABQuADAAS6EAsA3GZTYQHQAyAlpgI5OlmAZlYA2GvioBRTOjCIIMAPYIpAD3iYoXJVqoBvMlWM0RtcQBUIAazgA5OBplyFwADZwAPBYB8VK7YOGt4+ABQwABZcUFQAks7yCIoqIVRwsonJyFQ8MFQAxkpYMACUBkYmlbgA7FSYcADu/jb2jjAJrh4hoekuScrZhcUlApUAvmQTguQisBj5eQGtThmdXr7lpJUADghcAG6KcFQIcBAAJrpuAJ5xHf0pG9sQp5gwoya7B0c5bwVFbw+xgqxlwIiWQXaqyS7jgoXi0KyqWerxgSF+eSGbzKhi2lRMkWibBR6TyAF4qCTASD8YSoGwseT/sUgVQppUaWDmoE2vckRsAOJwKF9LKhHE0qq1epNCG8xEDbpUtHM7GsqZTYRUOboBbc5YizKKja4nZ7Q7wE5nS6YG53BWPPzK1lfC3HXKq940znglqQvkDeEBx2Ul6k9EexkSvH4qh04lhv4U52SglRemMqgUxnq70x0xUYBKJRuKgAWSU+wN4s2sdBtUZAFpG1Q/NQAGTt0OotgVquQ8W5/M+/xUADC6AQqKzTsTMDYE6npIEmvIhHEBFYtVNJgIkh3scOCBOEEwAHNjhS+bC2AAlU8X0KEdGECgjVPGI9pB3HilJc9wGwQqGgogbvvmHIQSYABmKhnPkERUKEX6YL836igMbByhooRCCUZS0AAnKEmDgeMkxAA)

**Comment by Kir-Antipov** [comment]
> No, that is not what you talk about in 3. HaloFour means that, with C# 9, you can write this extension method

@333fred, yes it is, it's not that different. We still have these cons:

 - It's not built-in feature, so we must reinvent this from project to project
 - Really strange logic: turn an enumerator into an ~~enumerable~~ object with `GetEnumerator`  so that foreach can get an enumerator from it...

**Comment by CyrusNajmabadi** [comment]
> It's not built-in feature, so we must reinvent this from project to project

Extremely uncommon things work that way. You can just create a simple nuget lib for this one and reference it from an the projects you need.

**Comment by Kir-Antipov** [comment]
> No allocations necessary

@HaloFour, yep, that's true, but the simplest way to achieve this will cause allocation. How often do developers you know (including you) choose the best way instead of the simplest and most obvious?

> This is an uncommon use

I disagree. It's **very** common use: do *thing0* with that part of enumeration, do *thing1* with this part of enumeration etc

> Turning an enumerator back into an enumerable is exactly what you're asking for here

Meh. Yes, but no. Into an enumerable, but not into `IEnumerable` (or any other object that supports `GetEnumerator`)

**Comment by CyrusNajmabadi** [comment]
> Really strange logic:

It doesn't seem strange to me...  It sounds like exactly what you're asking for. The ability to foreach over it. That's how the language foreachs. You provide this pattern and it works. That's why we designed y it this was tbh.

**Comment by 333fred** [comment]
> Let's imagine that we need to treat the first element of the enumeration differently than the rest.

I also think you're missing probably the simplest version of this (and with `Span<T>`, it's even no copy!):

```cs
DoWork(enumeration[0]);
foreach (var element in enumeration[1..])
{
    DoAnotherWork(element);
}
```

**Comment by Kir-Antipov** [comment]
@333fred, good luck to compile it:

```csharp
IEnumerable<int> enumeration = Enumerable.Range(0, 10);
DoWork(enumeration[0]);
foreach (var element in enumeration[1..])
{
    DoAnotherWork(element);
}
```

:)

**Comment by Kir-Antipov** [comment]
> Extremely uncommon things work that way

@CyrusNajmabadi, I don't think so. I need this literally in every real-world project...

> You can just create a simple nuget lib for this one and reference it from an the projects you need

I'm already upset enough that I've to include `System.Linq.Async` as a separate package in each project. We're moving in some strange direction: we have cool mechanisms, but in order to work with them into that cool way we need to connect 20 different packages from NuGet ¯\\_(ツ)_/¯
Will I live to see the day that we have to reference `System.DateTime` from NuGet?

**Comment by alrz** [comment]
>Will I live to see the day that we have to reference System.DateTime from NuGet?


You already do . If DateTime doesn't meet your needs, you should probably use NodaTime.

**Comment by Kir-Antipov** [comment]
@alrz, hah, true :)

**Comment by Kir-Antipov** [comment]
Okay, will change my example to `System.Decimal`. You need it, but it's not in *common* use. So let's move it to separate NuGet

**Comment by HaloFour** [comment]
> Okay, will change my example to `System.Decimal`.

I take it you don't work in the financial sector.  That type is _very_ commonly used.

**Comment by Kir-Antipov** [comment]
@HaloFour, note that I've italicized the word "common" :) 

I'm just talking about the fact that, ok, maybe you bypass the need for this functionality, but this doesn't mean that it ain't important for others. I use decimal more often than double or float, but I know **a lot** of people that never used it at all. Does it mean we need to move decimal to the separate package for "guys from the financial sector"?

**Comment by HaloFour** [comment]
@Kir-Antipov 

Nothing will be removed.  That will break existing projects.

There was a bit of cleaning house in the move from .NET Framework to .NET Core as .NET Framework accreted a _lot_ of types over the years.  But even that had to be measured against how many projects it would block from migrating.  Going forward the runtime teams are trying to avoid building everything into a single image and are less likely to accept API proposals that can be easily provided by an outside/NuGet library.

**Comment by CyrusNajmabadi** [comment]
> we need to connect 20 different packages from NuGet

And...?  That seems like it's working exactly as intended.  Not all functionality is core functionality.  When something is trivially solvable by a one line library function that is trivial to distribute, there's no need for a language change.

**Comment by Kir-Antipov** [comment]
> Nothing will be removed. That will break existing projects.

@HaloFour, yep, I know that perfectly.C# team agreed to apply the breaking change only once in my memory. It was a question from the category: "Let's imagine, that it won't. Would you?"

**Comment by Kir-Antipov** [comment]
> something is trivially solvable by a one line library function

@CyrusNajmabadi, one word, 3 TiB, int.MaxValue references and ulong.MaxValue vulnerabilities: `node_modules`

**Comment by CyrusNajmabadi** [comment]
> So let's move it to separate NuGet

If we were on 1.0 i would likely support that.  But we're  not :)  The dsicussion is not about moving exisiting apis to nuget packages, but instead how new language features aren't really needed if they are trivially provided by a nuget package.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi, one word, 3 TiB, int.MaxValue references and ulong.MaxValue vulnerabilities: node_modules

I don't know what any of that means :)  

What is 3TiB in reference to?  What do you mean by "int.MaxValue references?"  What would having this API be in a nuget package have to do with vulnerabilities?  I really don't understand what you're trying to say there.  Can you clarify?  Thanks!

**Comment by CyrusNajmabadi** [comment]
> I need this in literally every real-world project...

Common for you is not the litmus.  And, again, even if this were common for *everyone*, it's trivially solved for everyone with a simple function or nuget package.  So why update the language when the appropriate mechanism for this sort of thing is alreayd available and would work well here?

**Comment by Kir-Antipov** [comment]
> If we were on 1.0 i would likely support that. But we're not :)

@CyrusNajmabadi

![Understandable](https://user-images.githubusercontent.com/40232067/92634482-0b547500-f2dd-11ea-9484-7b8102e1fc5a.jpg)

**Comment by Kir-Antipov** [comment]
@CyrusNajmabadi, ok, I've got your views about this kind of subjects :)

**Comment by Kir-Antipov** [comment]
> I don't know what any of that means :)

@CyrusNajmabadi

I never liked how node_modules and these are the reasons:

- `3 TiB` - it's about node_modules size. A lot of developers implement and reimplement similar auxiliary functionality for their libraries, so that aggregated weight of their libraries increases, which wouldn't happen if the common functionality was built into the standard platform

- `int.MaxValue references` - there're a lot of "one one-line function" libraries

- `ulong.MaxValue vulnerabilities` - Who creates these packages? Who maintains them? When functionality is excessively fragmented into different packages, this only leads to an increase in the number of vulnerabilities among them

And it's kinda sad, that our ecosystem slowly turns into something node_modules like.

**Comment by HaloFour** [comment]
@Kir-Antipov 

> And it's kinda sad, that our ecosystem slowly turns into something node_modules like.

A healthy developer ecosystem is one where contributions are available from a lot of different sources.  Microsoft shouldn't be expected to ship every bit of functionality out of the default libraries.  Apart from it being a completely untenable strategy that will result in a mono-framework that is gigs in size it also precludes the use contributions from others.  In my opinion MS already ships too many frameworks and components.  I honestly think that Java has a larger and more vibrant ecosystem because a lot of larger organizations came together to fill in the many gaps left by Sun and Oracle.

**Comment by Kir-Antipov** [comment]
@HaloFour 

Let's start with the basics. There're two polar stages of ecosystems, both of which are completely unhealthy:
 
 - **Mono-framework** - it's built-in feature, or there's no such feature
 - **No-framework** - we don't have anything, but there's huge community and billions of packages for any need

I **am not** expecting .NET be "mono-framework" or "no-framework" either, cause they're equally terrible. 

For a long time .NET was 
![perfectly balanced.](https://user-images.githubusercontent.com/40232067/92638574-23c78e00-f2e3-11ea-9568-601ad9fae894.jpg)

.NET was the healthiest ecosystem I've ever know, and it's one of many things I love about it. For the first time
![I felt a great disturbance in the force](https://user-images.githubusercontent.com/40232067/92638925-c122c200-f2e3-11ea-91a4-96b4f8f64024.jpg)
was when we got IAsyncEnumerable, but the Linq methods for it - did not. And since then, every project I have has a reference to `System.Linq.Async`, and every output build comes out 1 MB heavier.

As I understood from **CyrusNajmabadi**'s words, we're moving to that side where our projects will be filled with hundreds of the references. And it disappoints me

**Comment by HaloFour** [comment]
@Kir-Antipov 

> and every output build comes out 1 MB heavier.

It would have always been 1 MB heavier.  The question is whether or not you've be able to choose to be 1 MB lighter in those projects where you don't need `System.Linq.Async`.  Being a separate package makes that possible.  Being embedded in the runtime does not.

**Comment by CyrusNajmabadi** [comment]
> it's about node_modules size. A lot of developers implement and reimplement similar auxiliary functionality for their libraries, 

So contribute PRs that reduce that and help the ecosystem solidify on canonical impls.

>  there're a lot of "one one-line function" libraries

Ok?

I don't get the arugment?  Should all one liners then get packaged into the core framework?  You're just trading off one thing for another.  Effectively an enormous class lib that attempts to pack in everything everyone needs for every situation.

> Who creates these packages? 

People who feel like there is value in the ecosystem having this function.

> Who maintains them? 

Those same people.

> When functionality is excessively fragmented into different packages, this only leads to an increase in the number of vulnerabilities among them

When functionality is consolidated, now the team has to spend all their time just maintaining millions of single line functions.  

You are stating we should take on this cost along with everything else, despite it being niche, and despite it being trivial for you to write and maintain yourself.  Now consider we have millions of customers.  *all* of them would love it if we did this for their helpers that they find value in.  Us taking on thousands or millions of those cases is a burden we cannot handle.  However, the ecosystem and community is vast and capable.  By spreading this cost out to all of you, it actually can work.

**Comment by CyrusNajmabadi** [comment]
@Kir-Antipov Please keep the memes out.  They're not helpful for this discussion.

**Comment by CyrusNajmabadi** [comment]
> For a long time .NET was

It really wasn't.  It was a split *you* found good for your needs.  But that doesn't apply to everyone.  There is always some group that wants their pet functionality in the core framework.  And every group will always find that some of what they want doesn't apply to the ecosystem as a whole.  the only maintainable approach is to have the core lib actually be subtantively valuable to the larger ecosystem as a whole, and have niche functionality be provided through third party libs.  What you're finding now is that this is a case of such niche functionality, and you are now on hte side that wants that.

**Comment by CyrusNajmabadi** [comment]
> As I understood from CyrusNajmabadi's words, we're moving to that side where our projects will be filled with hundreds of the references. And it disappoints me

If your projects have hundreds of niche needs, then yes, that is where you would end up.  However, such projects sound like they would be exceedingly niche themselves.  The vast majority of projects are not like that.  They generally get most of their functionality from teh core lib, and then get the rest from ecosystem libs.

**Comment by Kir-Antipov** [comment]
> Please keep the memes out. They're not helpful for this discussion.

@CyrusNajmabadi, I've already closed issue and discussion about the feature. And I just like speaking in memes, don't see anything criminal about it ¯\\_(ツ)_/¯

**Comment by Kir-Antipov** [comment]
> It would have always been 1 MB heavier. The question is whether or not you've be able to choose to be 1 MB lighter in those projects where you don't need System.Linq.Async. Being a separate package makes that possible. Being embedded in the runtime does not.

@HaloFour, erm, not. Well, I'm user. I downloaded 100 apps that use `System.Linq.Async`. It gives me +100 MB of their total weight. If `System.Linq.Async` was embedded in the runtime it would be hundred times less.

**Comment by CyrusNajmabadi** [comment]
> And I just like speaking in memes, don't see anything criminal about it ¯\(ツ)/¯

Understood.  But it's really not helpful or contributing to this discussion.  :)

**Comment by CyrusNajmabadi** [comment]
> It gives me +100 MB of their total weight. 

I'm not seeing an actual problem here.  You're describing this as if this is actually something which needs a solution.  But from my perspective there's nothing that needs to be done there.  If many different projects want to reference their own particular sets of different package references... what is the issue?  Why is that a bad thing?   

Furthermore, again this is all happening in the context of a request to change the language.   The nuget ecosystem working as intended and as expected isn't in any way a motivating case for me to now say "ok... everyone who has a one-off api need now can get that change into the language" :)

**Comment by Kir-Antipov** [comment]
> So contribute PRs that reduce that and help the ecosystem solidify on canonical impls

You're talking about some kind of utopia where everyone does everything right, or help those who did it wrong though. As a resident of a country that has spent almost a century figuring out whether utopias work, I can responsibly declare that they do not

> Ok?

It's not okay. If so many people have a need for a similar one-line functionality for built-in types that is needed in everyday use, then it must of course be represented by the built-in framework.

> Should all one liners then get packaged into the core framework?

Of course they do not. Some of them should, some of them should not. E.g., packages like MoreLinq should be presented as separate libraries, but it would be a real disaster if Linq should be referenced through NuGet.

> When functionality is consolidated, now the team has to spend all their time just maintaining millions of single line functions.

As I said before, there's no need to include everything into the core framework. It's just wrong. And yes, maintaining core framework is difficult task, but we trust core framework developers and there's really not so many chances for them to leave a vulnerability in the code due their skills and professionalism.

> You are stating we should take on this cost along with everything else, despite it being niche, and despite it being trivial for you to write and maintain yourself. Now consider we have millions of customers. all of them would love it if we did this for their helpers that they find value in. Us taking on thousands or millions of those cases is a burden we cannot handle. However, the ecosystem and community is vast and capable. By spreading this cost out to all of you, it actually can work.

No, I'm not stating it xD
It's great that developers can create their own libraries, put the helpers they use into packages and share them. But some things need to be built into the core framework. Somewhere there is a line that separates one from the other, and I'm honestly not ready to somehow formalize where exactly it is :)
Cause of that I may seem to you (a fan of the idea of ​​splitting everything into pluggable packages, as I understood) as someone who wants to cram everything into one place. But this isn't the case. I'm *somewhere* in between these two ideas :)

**Comment by CyrusNajmabadi** [comment]
> It's not okay. If so many people have a need for a similar one-line functionality for built-in types that is needed in everyday use, 

But not so many people have this need.  

> then it must of course be represented by the built-in framework.

It really does not.  Again, things need to be substantially valuable to make that bar.  Just like they need to be subtantially valuable to make it into a language feature.  A niche lang feature that is trivially solved with a one line niche API helper is absolutely *not* that. :)

**Comment by CyrusNajmabadi** [comment]
> but it would be a real disaster if Linq should be referenced through NuGet.

I honestly don't see why.  There are tons of useful, hugely used bits of functionality that come through nuget.  You seem to have a presupposed position that that is not ok.  

If i were doing Linq again today, I would likely fully support it, and all the Linq variants (like XLinq, etc.) come through nuget.  It would make a ton of sense.

**Comment by CyrusNajmabadi** [comment]
> Somewhere there is a line that separates one from the other, and I'm honestly not ready to somehow formalize where exactly it is :)

It's pretty simple: does this substantively improve things for the larger community out there.  The runtime team makes this determination all the time and it doesn't seem particularly hard to gauge :)

> I'm somewhere in between these two ideas :)

Sure.  IN this case though, the decision seems trivially clear.  This is def so niche that it should be external.  It def should not be in the language.  I can say that part at least for sure :)

**Comment by Kir-Antipov** [comment]
> pet functionality in the core framework

As I said before, this kind of things is bad. But `System.Linq.Async` **is not** "pet functionality", cause we have built-in IAsyncEnumerable, but we don't have built-in tools for it. And it's strange. I'd even be ok with the idea of the whole IAsyncEnumerable thing as a NuGet package, but this separation looks kinda strange: it's built-in, but it's not at the same time

**Comment by Kir-Antipov** [comment]
> A good way to think aobut this is that this is not a general fun-forum for random chats. It's an actual professional environment where people (including team members like myself) are engaging as part of our jobs and the work we do . Just as these memes wouldn't be appropriate as part of general emails on work topics, they're not appropriate here. So if you can please keep them to a minimum (or just elide entirely), that would be appreciated

I understood that, but personally, I don’t think that the workplace should always be something serious. Sometimes it can be "defused" with a joke or a meme. This is my opinion, but I respect those who don't think so and don't bother them with such things :)

**Comment by Kir-Antipov** [comment]
> I'm not seeing an actual problem here. You're describing this as if this is actually something which needs a solution. But from my perspective there's nothing that needs to be done there. If many different projects want to reference their own particular sets of different package references... what is the issue? Why is that a bad thing?

And I didn't get why it's not a bad thing xD Not everyone has few TB on their laptops, e.g. :) One day I had 128 GB in total, and something like this would be a great problem for me as a user.

**Comment by CyrusNajmabadi** [comment]
> One day I had 128 GB in total, and something like this would be a great problem for me as a user.

Sorry.  That's just not a use case we optimize the language for.  If your development environment cannot handle package sources, then that's not going to lead us to just put these in the language.  

To be as clear as possible: it's not htat this argument carries little weight.  It's explicitly that catering the language to this sort of case is a *non-goal*.  :)

**Comment by Kir-Antipov** [comment]
> If i were doing Linq again today, I would likely fully support it, and all the Linq variants (like XLinq, etc.) come through nuget. It would make a ton of sense

Just what I'm talking about. It's great for developers but it's not so great for users. Imagine that everyone ships their apps with the whole .NET (referenced part by part through NuGets) onboard. I really want to understand your point of view and why it's not bad

>  If your development environment cannot handle package sources, then that's not going to lead us to just put these in the language

I'm talking not about development, I'm talking about using that comes out of the development :)

**Comment by CyrusNajmabadi** [comment]
> Imagine that everyone ships their apps with the whole .NET (referenced part by part through NuGets) onboard.

Yes... i can imagine that.  It makes a whole lot of sense to me.  After all, i would want any particular app to be updatable independently from all other apps.  So having the runtime and deps be versionable and isolated in this fashion seems like a high positive for me :)

> I really want to understand your point of view and why it's not bad

Well, what would be bad about this approach?

>  I'm talking about using that comes out of the development :)

I would expect this approahc to, in general, decrease the size of any deployment due to not including lots of stuff you don't need, and only including the stuff you do.

**Comment by CyrusNajmabadi** [comment]
Note: at this point i think the discusison is better served in a venue like gitter.im/csharplang, or discord.gg/csharp .

**Comment by spydacarnage** [comment]
I remember a time when every application installed in Windows had a
collection of shared DLLs - each one tried to install it and if they had a
more up-to-date copy, it would overwrite the previous one, potentially
causing other applications to break.  And woe betide any time you tried to
uninstall something that was shared...

These days, these same applications have their own set of DLLs, not relying
on this shared space.  It involved a hit on installation sizes, on users'
hard drives, but I would never want to go back to how it was.

Each application having its own set of NuGet packages is just an extension
to that, and it allows multiple applications to use multiple versions of
the same NuGet package without any overriding or breaking of other
applications.

>

**Comment by Kir-Antipov** [comment]
> Well, what would be bad about this approach?

Application's size. If everyone ships there apps with the same libraries, users will spent in total much more space than if there was a place where all these binaries are gathered together. No? Or am I missing something?

> Note: at this point i think the discusison is better served in a venue like gitter.im/csharplang, or discord.gg/csharp .

Will keep in mind, but it's really last thing I want to know xD

**Comment by Kir-Antipov** [comment]
> These days, these same applications have their own set of DLLs, not relying
on this shared space.  It involved a hit on installation sizes, on users'
hard drives, but I would never want to go back to how it was.

@spydacarnage, why not to take goods from the both worlds? Cause "hit on installation sizes, on users' hard drives" is not something insignificant, we should care about users too.

**Comment by CyrusNajmabadi** [comment]
> , we should care about users too.

In one area you have dll hell, and it's enormously painful and almost impossible for either users, developers, or platforms to effectively solve.  In hte other, you have isolation, and extra space used.  It works well, and only comes at hte cost of HD space, a quantity that is both cheap and continuously dropping in cost.

Caring about users leads us to adapting to this and designing accordingly.  If HD space is cheap, plentiful, and continuing trends in that direction, then it's a better state to optimize for that, and help users avoid those enormously painful situations that we used to have before.

**Comment by CyrusNajmabadi** [comment]
> If everyone ships there apps with the same libraries, users will spent in

The issue is that apps do *not* ship the same libraries.  All apps use different versions of those libraries.  You cannot share because incompatibilities in libs will cause breaks in the different apps.

> Cause "hit on installation sizes, on users' hard drives" is not something insignificant, 

The hit there is vastly less significant or impactful than the hit caused by trying to share. With sharing we used to literally have apps simply stop functioning because a user upgraded another app.

**Comment by Kir-Antipov** [comment]
@CyrusNajmabadi, well that makes sense :)

**Comment by svick** [comment]
@Kir-Antipov 

> Imagine that everyone ships their apps with the whole .NET

That's completely normal with .Net Core, it's called [self-contained deployment](https://docs.microsoft.com/en-us/dotnet/core/deploying/#publish-self-contained).

**Comment by Kir-Antipov** [comment]
@svick, yep, I know (btw, read the first paragraph of the "Disadvantages" section), but I also know nobody who uses this for anything except loaded ASP.NET Core applications

**Comment by jnm2** [comment]
@Kir-Antipov Not true, this is great for desktop apps too.

**Comment by Kir-Antipov** [comment]
@jnm2, what's "not true"? That I don't know anybody who uses it? Sorry, didn't know you were me.

Guys, I've already said, that this whole thing makes sense to me. It doesn't mean that I agree with this (and I have reasons not to, but they go a little outside the scope of development-only environment, so I'm not trying to give them as arguments), this means that I do not reject it, so there's no need to convince me ¯\\_(ツ)_/¯

**Comment by jnm2** [comment]
Misread as "I know nobody uses this for anything" etc, I'm sorry.
